File: src/nostr/nip_44.rs
use base64::{engine::general_purpose, Engine as _};
use chacha20::cipher::StreamCipher;
use chacha20::ChaCha20;
use chacha20::cipher::KeyIvInit as _;
use hkdf::Hkdf;
use hmac::{Hmac, Mac};
use rand::RngCore;
use sha2::{Sha256};
use crate::nostr::{RelayError, Result};

pub fn encrypt_nip44(shared_secret: &[u8], plaintext: &str) -> Result<String> {
    let conversation_key = derive_conversation_key(shared_secret);
    let mut nonce = [0u8; 32];
    rand::thread_rng().fill_bytes(&mut nonce);

    let (chacha_key, chacha_nonce, hmac_key) = derive_message_keys(&conversation_key, &nonce);

    let padded = pad(plaintext);
    let mut ciphertext = padded.clone();
    let mut cipher = ChaCha20::new_from_slices(&chacha_key, &chacha_nonce)
        .map_err(|e| RelayError::CryptoError(e.to_string()))?;
    cipher.apply_keystream(&mut ciphertext);

    let mut hmac = Hmac::<Sha256>::new_from_slice(&hmac_key)
        .map_err(|e| RelayError::CryptoError(e.to_string()))?;
    hmac.update(&nonce);
    hmac.update(&ciphertext);
    let mac = hmac.finalize().into_bytes();

    let mut payload = Vec::with_capacity(1 + 32 + ciphertext.len() + 32);
    payload.push(0x02);
    payload.extend_from_slice(&nonce);
    payload.extend_from_slice(&ciphertext);
    payload.extend_from_slice(&mac);

    Ok(general_purpose::STANDARD.encode(payload))
}

pub fn decrypt_nip44(shared_secret: &[u8], encrypted_content: &str) -> Result<String> {
    if encrypted_content.len() > 87472 {
        return Err(RelayError::Generic("NIP-44 payload too large".into()));
    }

    let payload = general_purpose::STANDARD
        .decode(encrypted_content)
        .map_err(|e| RelayError::Base64Error(e.to_string()))?;

    if payload.len() > 65603 {
        return Err(RelayError::Generic("NIP-44 decoded payload too large".into()));
    }

    if payload.is_empty() || payload[0] != 0x02 {
        return Err(RelayError::Generic("Unsupported NIP-44 version".into()));
    }

    if payload.len() < 1 + 32 + 32 {
        return Err(RelayError::Generic("Invalid NIP-44 payload length".into()));
    }

    let nonce = &payload[1..33];
    let mac_start = payload.len() - 32;
    let ciphertext = &payload[33..mac_start];
    let mac = &payload[mac_start..];

    let conversation_key = derive_conversation_key(shared_secret);
    let (chacha_key, chacha_nonce, hmac_key) = derive_message_keys(&conversation_key, nonce);

    let mut hmac = Hmac::<Sha256>::new_from_slice(&hmac_key)
        .map_err(|e| RelayError::CryptoError(e.to_string()))?;
    hmac.update(nonce);
    hmac.update(ciphertext);
    if hmac.verify_slice(mac).is_err() {
        return Err(RelayError::CryptoError("Invalid NIP-44 MAC".into()));
    }

    let mut plaintext = ciphertext.to_vec();
    let mut cipher = ChaCha20::new_from_slices(&chacha_key, &chacha_nonce)
        .map_err(|e| RelayError::CryptoError(e.to_string()))?;
    cipher.apply_keystream(&mut plaintext);

    unpad(&plaintext)
}

fn derive_conversation_key(shared_secret: &[u8]) -> [u8; 32] {
    let (prk, _) = Hkdf::<Sha256>::extract(Some(b"nip44-v2"), shared_secret);
    let mut key = [0u8; 32];
    key.copy_from_slice(&prk);
    key
}

fn derive_message_keys(conversation_key: &[u8; 32], nonce: &[u8]) -> ([u8; 32], [u8; 12], [u8; 32]) {
    let hkdf = Hkdf::<Sha256>::from_prk(conversation_key).expect("HKDF from_prk failed");
    let mut okm = [0u8; 76]; // 32 + 12 + 32
    hkdf.expand(nonce, &mut okm).expect("HKDF expand failed");

    let mut chacha_key = [0u8; 32];
    let mut chacha_nonce = [0u8; 12];
    let mut hmac_key = [0u8; 32];

    chacha_key.copy_from_slice(&okm[0..32]);
    chacha_nonce.copy_from_slice(&okm[32..44]);
    hmac_key.copy_from_slice(&okm[44..76]);

    (chacha_key, chacha_nonce, hmac_key)
}

pub fn pad(plaintext: &str) -> Vec<u8> {
    let bytes = plaintext.as_bytes();
    let unpadded_len = bytes.len();
    
    let padded_len = if unpadded_len <= 32 {
        32
    } else {
        let next_power = 1 << (((unpadded_len - 1) as f64).log2().floor() as u32 + 1);
        let chunk = if next_power <= 256 {
            32
        } else {
            next_power / 8
        };
        chunk * ((unpadded_len - 1) / chunk + 1)
    };

    let mut padded = Vec::with_capacity(2 + padded_len);
    padded.extend_from_slice(&(unpadded_len as u16).to_be_bytes());
    padded.extend_from_slice(bytes);
    padded.resize(2 + padded_len, 0);
    padded
}

pub fn unpad(padded: &[u8]) -> Result<String> {
    if padded.len() < 2 {
        return Err(RelayError::Generic("Invalid padding".into()));
    }
    let len = u16::from_be_bytes([padded[0], padded[1]]) as usize;
    if len + 2 > padded.len() {
        return Err(RelayError::Generic("Invalid padding length".into()));
    }
    String::from_utf8(padded[2..2 + len].to_vec()).map_err(|e| RelayError::Utf8Error(e.to_string()))
}

#[cfg(test)]
#[path = "nip_44_test.rs"]
mod nip_44_test;

File: src/nostr/nip44.ts
import { sha256 } from '@noble/hashes/sha2.js';
import { hmac } from '@noble/hashes/hmac.js';
import { hkdf } from '@noble/hashes/hkdf.js';
import { chacha20 } from '@noble/ciphers/chacha.js';
import { RelayError } from './error.ts';
import { base64Encode, base64Decode } from '../util.ts';

function deriveConversationKey(sharedSecret: Uint8Array): Uint8Array {
  return hmac(sha256, new TextEncoder().encode('nip44-v2'), sharedSecret);
}

function deriveMessageKeys(conversationKey: Uint8Array, nonce: Uint8Array): { chachaKey: Uint8Array; chachaNonce: Uint8Array; hmacKey: Uint8Array } {
  const okm = hkdf(sha256, conversationKey, new Uint8Array(0), nonce, 76);
  return {
    chachaKey: okm.slice(0, 32),
    chachaNonce: okm.slice(32, 44),
    hmacKey: okm.slice(44, 76),
  };
}

export function pad(plaintext: string): Uint8Array {
  const bytes = new TextEncoder().encode(plaintext);
  const unpaddedLen = bytes.length;

  let paddedLen: number;
  if (unpaddedLen <= 32) {
    paddedLen = 32;
  } else {
    const nextPower = 1 << (Math.floor(Math.log2(unpaddedLen - 1)) + 1);
    const chunk = nextPower <= 256 ? 32 : nextPower / 8;
    paddedLen = chunk * Math.ceil(unpaddedLen / chunk);
  }

  const padded = new Uint8Array(2 + paddedLen);
  padded[0] = (unpaddedLen >> 8) & 0xff;
  padded[1] = unpaddedLen & 0xff;
  padded.set(bytes, 2);
  return padded;
}

export function unpad(padded: Uint8Array): string {
  if (padded.length < 2) throw RelayError.Generic('Invalid padding');
  const len = (padded[0] << 8) | padded[1];
  if (len + 2 > padded.length) throw RelayError.Generic('Invalid padding length');
  return new TextDecoder().decode(padded.slice(2, 2 + len));
}

export function encryptNip44(sharedSecret: Uint8Array, plaintext: string): string {
  const conversationKey = deriveConversationKey(sharedSecret);
  const nonce = crypto.getRandomValues(new Uint8Array(32));
  const { chachaKey, chachaNonce, hmacKey } = deriveMessageKeys(conversationKey, nonce);

  const padded = pad(plaintext);
  const ciphertext = chacha20(chachaKey, chachaNonce, padded);

  const mac = hmac(sha256, hmacKey, new Uint8Array([...nonce, ...ciphertext]));

  const payload = new Uint8Array(1 + 32 + ciphertext.length + 32);
  payload[0] = 0x02;
  payload.set(nonce, 1);
  payload.set(ciphertext, 33);
  payload.set(mac, 33 + ciphertext.length);

  return base64Encode(payload);
}

export function decryptNip44(sharedSecret: Uint8Array, encryptedContent: string): string {
  if (encryptedContent.length > 87472) {
    throw RelayError.Generic('NIP-44 payload too large');
  }

  const payload = base64Decode(encryptedContent);

  if (payload.length > 65603) {
    throw RelayError.Generic('NIP-44 decoded payload too large');
  }

  if (payload.length === 0 || payload[0] !== 0x02) {
    throw RelayError.Generic('Unsupported NIP-44 version');
  }

  if (payload.length < 1 + 32 + 32) {
    throw RelayError.Generic('Invalid NIP-44 payload length');
  }

  const nonce = payload.slice(1, 33);
  const macStart = payload.length - 32;
  const ciphertext = payload.slice(33, macStart);
  const mac = payload.slice(macStart);

  const conversationKey = deriveConversationKey(sharedSecret);
  const { chachaKey, chachaNonce, hmacKey } = deriveMessageKeys(conversationKey, nonce);

  const expectedMac = hmac(sha256, hmacKey, new Uint8Array([...nonce, ...ciphertext]));
  if (!constantTimeEqual(mac, expectedMac)) {
    throw RelayError.CryptoError('Invalid NIP-44 MAC');
  }

  const plaintext = chacha20(chachaKey, chachaNonce, ciphertext);

  return unpad(plaintext);
}

function constantTimeEqual(a: Uint8Array, b: Uint8Array): boolean {
  if (a.length !== b.length) return false;
  let diff = 0;
  for (let i = 0; i < a.length; i++) {
    diff |= a[i] ^ b[i];
  }
  return diff === 0;
}
